### ISSUE TITLE
Allow use of DML with RETURNING clause in SELECT

### ISSUE THREAD

**Issue by mrotteveel** [FirebirdSQL/firebird#7573]
It would be useful to allow access to DML with `RETURNING` clauses in `SELECT`. The SQL standard itself provides something called a _"table function derived table "_ (see 7.6 \<table reference\> in SQL:2016-2), allowing you to access a table-valued function as a table. Given DML with `RETURNING` could be considered a table-valued function, this syntax could be useful to provide access to DML in select.

Note that SQL standard also has something called a _"data change delta table"_ (see also section 7.6 in SQL:2016-2), which is intended to do the same (providing access to the changes of a DML statement in a SELECT), but it comes with additional clauses which makes no sense for us, but essentially below proposal is a combination of _"table function derived table "_ and _"data change delta table"_.

My proposal is the follows:

```
<data-change-table> ::= TABLE '(' <modifying-dml> ')' [[AS] <alias> [(<derived-column-list>)]]
<modifying-dml> :: = {<insert> | <update-searched> | <update-or-insert> | <delete-searched> | merge }  

<table-primary> ::= { ..existing options.. | <data-change-table> }
```
With the added requirement that the modifying-dml **must** have a `RETURNING` clause. The data-change-table would be similar to a derived-table.

It should be possible to use columns of joined tables as parameters, similar to using selectable stored procedures, but we may need to consider if that should require explicit use of `LATERAL` or not.

We could also consider extending the same treatment to `EXECUTE BLOCK` with `RETURNS`, and maybe `EXECUTE PROCEDURE`, but that might be something for later.

Alternatives:
* The requirement that a statement must have `RETURNING` could also be replaced with providing an empty table stream if the `RETURNING` clause is absent. This might not seem very useful, but will allow calling a DML statement from within a select, even if you're not interested in the result.

**Comment by asfernandes** [comment]
I don't understand why you want to SELECT have RETURNING.
Isn't RETURNING the same as the SELECT value list?
Please give some usage examples.

**Comment by mrotteveel** [comment]
I'm not talking about adding a `RETURNING` to `SELECT`, I'm talking about allowing you to use an INSERT, UPDATE, UPDATE OR INSERT, DELETE or MERGE with a `RETURNING` clause as a form of derived-table in a select statement.

So for example:
```sql
select i.id, o.some_name, i.some_value
from table (
  insert into sometable (id, some_id, some_value) values (1, 513, 'abc')
  returning id, some_id, some_value
) i
inner join othertable o
  on o.some_id = i.some_id
```

**Comment by mrotteveel** [comment]
A bit more complicated example for 'archiving' data to another table (using `LATERAL` here, but we may not need that)

```sql
select o.id as original_id, n.id as archive_id, l.id updated_link_id
from table (
  delete from original_table where expiry_date < current_date 
  returning id, some_data_1, some_data_2, some_data_3
) o,
lateral table (
  insert into archive_table (some_data_1, some_data_2, some_data_3) 
    values (o.some_data_1, o.some_data_2, o.some_data_3)
  returning id
) n,
lateral table (
  update some_linked_table set archive_id = n.id where original_id = o.id
  return id
) l
```

**Comment by tonne1** [comment]
+ 1: this could be quite useful, actually just tested this yesterday to see if was possible.
Could also be a workaround for https://github.com/FirebirdSQL/firebird/issues/4722

**Comment by mrotteveel** [comment]
PostgreSQL actually does support something like this:

```
with a as (
   insert into example(val) values ('abc') returning id, val
)
select * 
from a
left join (values ('abc', 'x')) x (lookup, val2)
  on a.val = x.lookup
```

See also [Data-Modifying Statements in WITH](https://www.postgresql.org/docs/15/queries-with.html#QUERIES-WITH-MODIFYING)

Though my suggestion with allowing it to use as lateral tables to do additional work based on previous tables is not supported in PostgreSQL as far as I can tell.

**Comment by mrotteveel** [comment]
> I am categorically against such constructions. Do not forget that the optimizer can choose any order of joining tables, which makes the result of such queries unpredictable.

1) You can change the optimizer
2) Lateral tables establish a dependency order based on the correlated values, so the order is already deterministic
3) For uncorrelated modifying DML statements, you can execute the DML **once** and use the resulting table as a stable record source (this is basically what PostgreSQL does with its `WITH` solution).
4) Maybe the lateral solution for correlated modifying DML is a bridge too far right now, but as a first step we could support uncorrelated DML this way, and refine/improve it later to allow correlated DML

**Comment by sim1984** [comment]
By the way, it would be possible to use internal local temporary tables, which are now used for returning to implement queries like 'values (1,2), (3, 4)...'

**Comment by aafemt** [comment]
> Do not forget that the optimizer can choose any order of joining tables

Not quite so. Returning table is not indexed in any way so it hardly can be second stream. And even if plan is JOIN (A NATURAL, B NATURAL) the second stream is fully scanned.

**Comment by sim1984** [comment]
> 
Don't think in the past. Even for an unindexed table join, there are three ways to perform nested loop, hash join, merge join. The basic principle that should be followed is "we should not make any assumptions about the order of execution in an sql query."

**Comment by aafemt** [comment]
All of them require full read of the record stream for building hash-table, sorting or just make sure that there is no records left. And in any case "returning" that forms the record set is guaranteed by current implementation (which fills LTT) to finish all modifications before returning any result.

**Comment by mrotteveel** [comment]
That depends on the rules we establish for this construct.

I'd say:

1. For non-correlated `TABLE (<modifying-dml>)` it is **always** executed
2. For correlated (LATERAL) `TABLE (<modifying-dml>)`, the engine enforces a simple JOIN execution order (left to right, top to bottom), and if that is inefficient, you have only yourself to blame.

**Comment by sim1984** [comment]
Imagine that you are inserting records into 2 tables linked by a foreign key. In a complex query, there is no guarantee that the inserts will be done in the correct order. Moreover, when changing the distribution of records in other tables participating in queries, the order of execution may change.

In addition, consider a simple query

SELECT*
FROM table1
JOIN (INSERT INTO table2 ... RETURNING ...) ON ...

In addition, if for some reason a NESTED LOOP plan is chosen instead of a HASH / MERGE JOIN, then the INSERT will start executing repeatedly. In principle, this can probably be avoided, but as for me, at the moment it is dangerous to add a function that will work unstably.

But the ability to use FOR ... RETUNINING in PSQL would be useful and generally safe and predictable.

The table-value constructor is also a very useful and predictable feature.

**Comment by mrotteveel** [comment]
> Imagine that you are inserting records into 2 tables linked by a foreign key. In a complex query, there is no guarantee that the inserts will be done in the correct order. Moreover, when changing the distribution of records in other tables participating in queries, the order of execution may change.

Why? Such situations can be prevented by establishing rules about order for execution.

> In addition, consider a simple query
> 
> SELECT* FROM table1 JOIN (INSERT INTO table2 ... RETURNING ...) ON ...
> 
> In addition, if for some reason a NESTED LOOP plan is chosen instead of a HASH / MERGE JOIN, then the INSERT will start executing repeatedly. In principle, this can probably be avoided, but as for me, at the moment it is dangerous to add a function that will work unstably.

Again, that can be prevented by establishing rules about order of execution. You're extrapolating behaviour based on current optimizer behaviour when accessing normal tables as an argument to reject a feature, while those behaviours will not necessarily apply when implementing a new feature like this.

That said, preventing such behaviours may increase the complexity of implementing this.

> But the ability to use FOR ... RETUNINING in PSQL would be useful and generally safe and predictable.
> 
> The table-value constructor is also a very useful and predictable feature.

Although I'd really love to have table-value constructors (and if I have to prioritize, I'll choose table-value constructors over this), they will not give you access to things like generated identity values, computed column values, default values, trigger generated values, etc., while this feature will.

**Comment by aafemt** [comment]
> if for some reason a NESTED LOOP plan is chosen instead of a HASH / MERGE JOIN, 
> then the INSERT will start executing repeatedly.

No, it will repeatedly feed records from local temporary table formed once.

**Comment by asfernandes** [comment]
This feature seems for me, at start, very great, but as I thought on it more and more I tend to not support it.
It would have many rules to just understand how it works, and it covers an area that may be simple done already with EXECUTE BLOCK or routines. It would also be a nightmare to implement correctly.

**Comment by aafemt** [comment]
> It would have many rules to just understand how it works

No difference from monitoring snapshot in this regard.

**Comment by sim1984** [comment]
This is a complex and ambiguous function. Its implementation should at least be postponed after release of v5.

I propose to start implementing a simpler and more unambiguous feature

``` 
FOR
   <DML with RETURNING>
   INTO...
DO...
```

**Comment by mrotteveel** [comment]
> This is a complex and ambiguous function. Its implementation should at least be postponed after release of v5.
> 
> I propose to start implementing a simpler and more unambiguous feature
> 
> ```
> FOR
>    <DML with RETURNING>
>    INTO...
> DO...
> ```

That is not directly related to this feature proposal, which is a DSQL proposal. I just mentioned that it would be possible to use my proposed syntax as an alternative to having to implement `FOR <modifying-DML>`. What you want is tracked separately by #6925
